src/profile-logic/profile-data.js

```diff
@@ -594,46 +595,156 @@ export function filterThreadToRange(
+    return result;
+  }
+
+  // This map is basically the result of:
+  // `map.set(index, hashPath(callNodePath.slice(0, index)))`
+  // This is a cache to speed up contributing data to the `cache` map.
+  const sliceHashes = new Map();
+  sliceHashes.set(callNodePath.length, hashFullPath);
+
+  let i = callNodePath.length;
+  let index;
+  while (--i > 0) {
+    // Looking up each parent for this call node, starting from the deepest node.
+    // If we find a parent this makes it possible to start the look up from this location.
+    const subPath = callNodePath.slice(0, i);
+    const hash = hashPath(subPath);
+    index = cache.get(hash);
+    if (index !== undefined) {
+      // Yay, we already have the result for a parent!
+      break;
+    }
+    // Cache the hashed value because we'll need it later.
+    sliceHashes.set(i, hash);
+  }
+
+  result = _getCallNodeIndexFromPath(
+    callNodePath,
+    callNodeTable,
+    index,
+    i,
+    // This function is called at each "step" of the algorithm, so that we can
+    // cache the result for each node to speed up future lookups.
+    (callNodeIndex, indexInCallNodePath) => {
+      // indexInCallNodePath points to the func index that got resolved. That's
+      // why we need this `+ 1` incrementation to get the slice including this
+      // index.
+      // Reminder: this is the cached result of:
+      // `hashPath(callNodePath.slice(0, indexInCallNodePath + 1))`
+      const hash = sliceHashes.get(indexInCallNodePath + 1);
+      if (hash !== undefined) {
+        // caching the result for this path.
+        cache.set(hash, callNodeIndex);
+      }
+    }
+  );
+  return result;
+}
+
+// Returns the CallNodeIndex that matches the function `func` and whose parent's
+// CallNodeIndex is `parent`.
+function _getCallNodeIndexFromParentAndFunc(
+  parent: IndexIntoCallNodeTable,
+  func: IndexIntoFuncTable,
   callNodeTable: CallNodeTable
 ): IndexIntoCallNodeTable | null {
-  let fs = -1;
-  for (let i = 0; i < callNodePath.length; i++) {
-    const func = callNodePath[i];
-    let nextFS = -1;
-    for (
-      let callNodeIndex = fs + 1;
-      callNodeIndex < callNodeTable.length;
-      callNodeIndex++
+  for (
+    let callNodeIndex = parent + 1; // the root parent is -1
+    callNodeIndex < callNodeTable.length;
+    callNodeIndex++
+  ) {
+    if (
+      callNodeTable.prefix[callNodeIndex] === parent &&
+      callNodeTable.func[callNodeIndex] === func
     ) {
-      if (
-        callNodeTable.prefix[callNodeIndex] === fs &&
-        callNodeTable.func[callNodeIndex] === func
-      ) {
-        nextFS = callNodeIndex;
-        break;
-      }
+      return callNodeIndex;
     }
-    if (nextFS === -1) {
+  }
+
+  return null;
+}
+
+/*
+ * This function returns a CallNodeIndex from a CallNodePath, using the
+ * specified `callNodeTable`.
+ * The 2 additional arguments make it possible to start the algorithm with a
+ * specific state.
+ */
+function _getCallNodeIndexFromPath(
```

Comment:
I don't feel like we benefit from having this generic separate function. I think it would be better to inline it into the two callers.

Reply:
But I think I understand what you mean. I initially mostly wanted to be able to start the same algorithm as before, but at a specific step. And only then came the need for the callback. Maybe it was overcomplexified.